framework/src/actions/ResolveOptionalMaterialPropertiesAction.C

```diff
@@ -0,0 +1,35 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ResolveOptionalMaterialPropertiesAction.h"
+#include "FEProblem.h"
+
+registerMooseAction("MooseApp",
+                    ResolveOptionalMaterialPropertiesAction,
+                    "resolve_optional_materials");
+
+InputParameters
+ResolveOptionalMaterialPropertiesAction::validParams()
+{
+  InputParameters params = Action::validParams();
+  return params;
+}
+
+ResolveOptionalMaterialPropertiesAction::ResolveOptionalMaterialPropertiesAction(
+    InputParameters params)
+  : Action(params)
+{
+}
+
+void
+ResolveOptionalMaterialPropertiesAction::act()
+{
+  mooseAssert(_problem, "Problem doesn't exist");
+  _problem->resolveOptionalMaterialProperties();
```

Comment:
Yeah I think I would prefer looping through all of the materials from the warehouse here, but I'm not bent on it

Reply:
That was somewhat similar to my first design. Let me think about it. Each material would have it's own vector of proxy objects, and we would doubly nested loop over them in all materials. Yeah, I think that could work. That loop logic would be contained in the resolve properties action. And at the same time I'd get rid of the circular inclusion issue that the use of FEProblemBase brings with it. I like that.

Reply:
👍🏼 I'm all for making things more local when possible as long as it doesn't make it too ugly

Reply:
Done!